File: sdk/flutter/example/quickstart.dart
// Quickstart against a local dev node:  ./target/release/ferrum-node --dev
//
//   dart run example/quickstart.dart
//
// Anchors a DID and files a tax obligation as //Alice (the dev sudo/governance
// key — also accredit Alice as an issuer once via a sudo-wrapped registerIssuer).

import 'dart:typed_data';

import 'package:ferrum_sdk/ferrum_sdk.dart';
import 'package:pointycastle/digests/blake2b.dart';

String commit(String s) {
  final d = Blake2bDigest(null, 32);
  final bytes = Uint8List.fromList(s.codeUnits);
  d.update(bytes, 0, bytes.length);
  final out = Uint8List(32);
  d.doFinal(out, 0);
  return '0x${hexEncode(out)}';
}

Future<void> main() async {
  final ferrum = await FerrumClient.connect('ws://127.0.0.1:9944');
  final alice = await FerrumClient.keypair('//Alice');

  final subject = Did.of('tw', 'citizen-0001');

  final anchor = ferrum.identity.anchorDid(DidDocument(
    subject,
    Uint8List.fromList(alice.publicKey.bytes),
    commit('off-chain DID document for citizen #1'), // commitment only — no PII
    [DidKeyRef(KeyKind.sr25519, commit('device-key'))],
    commit('rev-acc-0'),
    0,
  ));
  print('anchor_did hash: ${await ferrum.signAndSend(anchor, alice)}');

  final obligation = ferrum.tax.fileObligation(TaxObligation(
    subject,
    TaxKind.income,
    FiatAmount('TWD', BigInt.from(1234500)),
    commit('encrypted return detail'),
    false,
  ));
  print('file_obligation hash: ${await ferrum.signAndSend(obligation, alice)}');

  await ferrum.disconnect();
}

File: sdk/rust/examples/quickstart.rs
//! Quickstart against a local dev node:  ./target/release/ferrum-node --dev
//!
//!     cargo run --example quickstart
//!
//! Accredits Alice as an issuer (via sudo), anchors a DID, files a tax obligation.

use blake2::{Blake2b, Digest};
use blake2::digest::consts::U32;
use ferrum_sdk::{dev, types::*, FerrumClient};
use subxt::utils::AccountId32;

fn commit(s: &str) -> [u8; 32] {
    let mut h = Blake2b::<U32>::new();
    h.update(s.as_bytes());
    h.finalize().into()
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let ferrum = FerrumClient::connect("ws://127.0.0.1:9944").await?;
    let alice = dev::alice();
    let alice_id: AccountId32 = alice.public_key().to_account_id();

    let subject = Did::new("tw", "citizen-0001");

    // 1. Governance accredits Alice as an issuer (sudo wraps governance on --dev).
    //    Nest the inner call as a RuntimeCall value: outer variant = pallet name,
    //    inner variant = call name + named fields.
    use subxt::ext::scale_value::Value;
    let pallet_call = Value::named_variant("register_issuer", [("who".into(), account(&alice_id))]);
    let runtime_call = Value::unnamed_variant("Identity", [pallet_call]);
    let sudo = subxt::dynamic::tx("Sudo", "sudo", vec![runtime_call]);
    ferrum.sign_and_send(&sudo, &alice).await?;
    println!("issuer accredited");

    // 2. Anchor the DID — doc_hash is a commitment, never the document itself.
    let doc = DidDocument {
        did: subject.clone(),
        controller: alice_id.clone(),
        doc_hash: commit("off-chain DID document for citizen #1"),
        keys: vec![DidKeyRef { kind: "Sr25519".into(), key_hash: commit("device-key") }],
        revocation_commitment: commit("rev-acc-0"),
        anchored_at: 0,
    };
    ferrum.sign_and_send(&ferrum.identity().anchor_did(&doc), &alice).await?;
    println!("DID anchored");

    // 3. File a fiat-denominated tax obligation.
    let obligation = TaxObligation {
        subject: subject.clone(),
        kind: "Income".into(),
        amount_due: FiatAmount { currency: "TWD".into(), minor_units: 1_234_500 },
        detail_commitment: commit("encrypted return detail"),
        settled: false,
    };
    ferrum.sign_and_send(&ferrum.tax().file_obligation(&obligation), &alice).await?;
    println!("obligation filed");

    Ok(())
}
